### PR TITLE
Enable datatree * dataset commutativity

### PR DESCRIPTION
**Pull request by TomNicholas**
Change binary operations involving `DataTree` objects and `Dataset` objects to be handled by the `DataTree` class. Necessary to enable `ds * dt` to return the same type as `dt * ds`.

Builds on top of #7418. 

- [x] Closes https://github.com/xarray-contrib/datatree/issues/146
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: doc/user-guide/io.rst [hunk 1]
```diff
@@ -228,10 +231,34 @@ Either of these groups can be loaded from the file as an independent :py:class:`
...
+If installed, this package's API can be imported directly from xarray, i.e. ``from xarray import DataTree``.
+
+Whilst netCDF groups can only be loaded individually as Dataset objects, a whole file of many nested groups can be loaded
+as a single :py:class:`DataTree` object.
+To open a whole netCDF file as a tree of groups use the :py:func:`open_datatree()` function.
+To save a DataTree object as a netCDF file containing many groups, use the :py:meth:`DataTree.to_netcdf()`` method.
+
+.. _netcdf.group.warning:
+
+.. warning::
```
Comment (shoyer) [inline doc/user-guide/io.rst:250]: We have this same issue with round-tripping Dataset objects. Unless it's particularly more pronounced for Datatree objects, I would consider consolidating the discussion, maybe in a new section like "netCDF files that Xarray cannot represent."

File: doc/user-guide/io.rst [hunk 2]
```diff
@@ -633,6 +660,21 @@ To read back a zarr dataset that has been created this way, we use the
...
+Like for netCDF, zarr groups can either be opened as individual :py:class:`Dataset` objects using the ``group`` keyword argument to :py:func:`open_dataset`,
+or alternatively nested groups in zarr stores can be represented by loading the store as a :py:class:`DataTree` object.
+(The latter option requires that you have the `xarray-datatree <https://github.com/xarray-contrib/datatree>`_ package installed.)
+
+To open a whole zarr store as a tree of groups use the :py:func:`open_datatree()` function.
+To save a DataTree object as a zarr store containing many groups, use the :py:meth:`DataTree.to_zarr()` method.
+
+.. note::
+    Note that perfect round-tripping should always be possible with a zarr store (:ref:`unlike for netCDF files<netcdf.group.warning>`),
+    as zarr does not support "unused" dimensions.
```
Comment (shoyer) [inline doc/user-guide/io.rst:675]: I would just leave this note under the netCDF section -- only people concerned about netCDF need to worry about this

File: xarray/core/dataarray.py [hunk 3]
```diff
@@ -3656,6 +3656,48 @@ def reduce(
...
+        """
+
+        try:
+            from datatree import DataTree
+        except ImportError:
+            raise ImportError(
+                "Could not import the datatree package. "
+                "Find it at https://github.com/xarray-contrib/datatree"
+            )
+
```
Comment (shoyer) [inline xarray/core/dataarray.py:3697]: Python will give more informative error message ("The above exception was the direct cause of the following exception") if you use `raise ... from ...` here:
```suggestion
        except ImportError as e:
            raise ImportError(
                "Could not import the datatree package. "
                "Find it at https://github.com/xarray-contrib/datatree"
            ) from e
```

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -6589,6 +6628,14 @@ def _binary_op(self, other, f, reflexive=False, join=None) -> Dataset:
         from xarray.core.dataarray import DataArray
         from xarray.core.groupby import GroupBy
 
+        try:
+            from datatree import DataTree
+
+            if isinstance(other, DataTree):
+                return NotImplemented
+        except ImportError:
+            pass
```
Comment (shoyer) [inline xarray/core/dataset.py:6637]: Consider using `contextlib.suppress()` here: https://docs.python.org/3/library/contextlib.html#contextlib.suppress

File: xarray/tests/test_datatree.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    other_ds = Dataset({"z": ("z", [0.1, 0.2])})
+
+    expected = DataTree(data=ds1 * other_ds)
+    DataTree(name="subnode", data=ds2 * other_ds, parent=expected)
+
+    result = dt * other_ds
+    dtt.assert_equal(result, expected)
+
+    # This ordering won't work unless xarray.Dataset defers to DataTree.
+    # See https://github.com/xarray-contrib/datatree/issues/146
```
Comment (shoyer) [inline xarray/tests/test_datatree.py:52]: Presumably this comment can be deleted now :)

File: xarray/tests/test_datatree.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+@requires_datatree
+def test_binary_ops():
+    import datatree.testing as dtt
+
+    from xarray import DataTree
+
+    ds1 = Dataset({"a": [5], "b": [3]})
+    ds2 = Dataset({"x": [0.1, 0.2], "y": [10, 20]})
+    dt = DataTree(data=ds1)
+    DataTree(name="subnode", data=ds2, parent=dt)
```
Comment (shoyer) [inline xarray/tests/test_datatree.py:42]: Does this line do anything?

File: xarray/tests/test_datatree.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    other_ds = Dataset({"z": ("z", [0.1, 0.2])})
+
+    expected = DataTree(data=ds1 * other_ds)
+    DataTree(name="subnode", data=ds2 * other_ds, parent=expected)
```
Comment (shoyer) [inline xarray/tests/test_datatree.py:46]: Same concern as above

**Review by shoyer — COMMENTED** [review]
Looking great, thank you!

**Comment by TomNicholas** [comment]
This was superceded by https://github.com/pydata/xarray/pull/9619

**Pull request closed without merging**